plugin/trino-hudi/src/main/java/io/trino/plugin/hudi/split/HudiBackgroundSplitLoader.java

```diff
@@ -0,0 +1,120 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package io.trino.plugin.hudi.split;
+
+import com.google.common.util.concurrent.FutureCallback;
+import com.google.common.util.concurrent.Futures;
+import com.google.common.util.concurrent.ListenableFuture;
+import io.airlift.concurrent.MoreFutures;
+import io.trino.plugin.hive.HivePartitionKey;
+import io.trino.plugin.hive.util.AsyncQueue;
+import io.trino.plugin.hudi.HudiTableHandle;
+import io.trino.plugin.hudi.partition.HudiPartitionInfo;
+import io.trino.plugin.hudi.partition.HudiPartitionInfoLoader;
+import io.trino.plugin.hudi.query.HudiFileLister;
+import io.trino.spi.connector.ConnectorSession;
+import io.trino.spi.connector.ConnectorSplit;
+import org.apache.hadoop.fs.FileStatus;
+import org.apache.hadoop.fs.FileSystem;
+
+import java.util.ArrayDeque;
+import java.util.Collection;
+import java.util.List;
+import java.util.concurrent.ExecutorService;
+import java.util.function.Consumer;
+import java.util.stream.Collectors;
+
+import static com.google.common.util.concurrent.MoreExecutors.directExecutor;
+import static java.util.Objects.requireNonNull;
+
+public class HudiBackgroundSplitLoader
+{
+    private final ConnectorSession session;
+    private final HudiFileLister hudiFileLister;
+    private final AsyncQueue<ConnectorSplit> asyncQueue;
+    private final ExecutorService executor;
+    private final Consumer<Throwable> errorListener;
```

Comment:
Heads'up: There is also in Hive `io.trino.plugin.hive.BackgroundHiveSplitLoader`.
Did you take inspiration from this class during the development?